andersoncd/datafits.py

```diff
@@ -24,4 +24,8 @@ def gradient(self, X, y, w, Xw):
         return (X.T @ Xw - self.Xty) / len(Xw)
 
     def gradient_scalar(self, X, w, Xw, j):
+        # TODO remove w unused in argument
         return (X[:, j] @ Xw - self.Xty[j]) / len(Xw)
+
+    def gradient_scalar_sparse(self, Xj, idx_nz, Xw, y):
+        return (Xj @ (Xw[idx_nz] - y[idx_nz])) / len(Xw)
```

Comment:
is this not inefficient as you build the array Xw[...] ? rather than looping for startptr to endptr

Reply:
also use Xty